# **Import and load the dataset**

In [9]:
import pandas as pd

from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split, KFold, cross_val_score
from sklearn.linear_model import LogisticRegression

# load the dataset
data = load_breast_cancer()

# create predictors and target
X = pd.DataFrame(data.data, columns=data.feature_names)
y = data.target

print("Total records:", len(X))

Total records: 569


# **Basic split**

In [10]:
# split the data into 80% training and 20% testing
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42,
    stratify=y
)

print("Training records:", len(X_train))
print("Testing records:", len(X_test))

Training records: 455
Testing records: 114


**split justification**

used an 80/20 split.

80% of the data is used for training and 20% is reserved for testing.
since the dataset contains only 569 records, this provides 455 training
records for the model while still keeping 114 independent records for testing.

stratify=y is also useful because it keeps approximately the same proportion of cancer classes in both sets.

i also used random_state=42 so the split can be reproduced.

# **K-Fold configuration**

In [11]:
# create 5 folds and shuffle the data before splitting
kfold = KFold(
    n_splits=5,
    shuffle=True,
    random_state=42
)

**K justification**

i selected k=5.

using 5 folds gives the model five different validation experiments while
keeping enough observations inside each fold. It provides a good balance
between reliability and computational cost.

a very large value of k would require training the model many more times,
increasing the computational cost significantly.

in this experiment, cross-validation is performed only on the 455 training
records, so k=569 would not be possible because the number of folds cannot
be greater than the number of training samples.

# **Logistic Regression + Cross-Validation**

In [12]:
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline

# create the model with data scaling
model = make_pipeline(
    StandardScaler(),
    LogisticRegression(max_iter=3000)
)

# evaluate the model using cross-validation
scores = cross_val_score(
    model,
    X_train,
    y_train,
    cv=kfold,
    scoring="accuracy"
)

# show the results
print("Fold accuracies:", scores)
print("Mean accuracy:", scores.mean())
print("Standard deviation:", scores.std())

Fold accuracies: [0.98901099 0.95604396 1.         0.97802198 0.93406593]
Mean accuracy: 0.9714285714285715
Standard deviation: 0.023671054097294517


**analysis**

The model appears to be relatively stable because the standard deviation
of the accuracy scores is low.

This means that the model obtains similar accuracy across the different
folds, so its performance does not depend heavily on one specific division
of the training data.

Therefore, the cross-validation results suggest that the model is stable
across different data splits.